### PR TITLE
Error handling cleanup in solvers/flattening 4

### PR DESCRIPTION
**Pull request by danpoe**

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 1161a36eafd4759cdd0346267cb0aa31f62fb9ac).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84460365

File: src/solvers/flattening/boolbv_replication.cpp [hunk 1]
```diff
@@ -18,29 +18,33 @@ bvt boolbvt::convert_replication(const replication_exprt &expr)
     return conversion_failed(expr);
 
   mp_integer times;
-  if(to_integer(expr.op0(), times))
-    throw "replication takes constant as first parameter";
+  bool error = to_integer(expr.op0(), times);
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/flattening/boolbv_replication.cpp]: I'd just replace this with `numeric_cast_v`

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]

File: src/solvers/flattening/boolbv_shift.cpp [hunk 2]
```diff
@@ -48,15 +45,17 @@ bvt boolbvt::convert_shift(const binary_exprt &expr)
...
+    UNREACHABLE;
 
   // we allow a constant as shift distance
 
   if(expr.op1().is_constant())
   {
     mp_integer i;
-    if(to_integer(expr.op1(), i))
-      throw "convert_shift: failed to convert constant";
+    bool error = to_integer(expr.op1(), i);
```
Comment (peterschrammel) [inline src/solvers/flattening/boolbv_shift.cpp]: const

File: src/solvers/flattening/boolbv_typecast.cpp [hunk 3]
```diff
@@ -590,20 +610,18 @@ bool boolbvt::type_conversion(
...
 literalt boolbvt::convert_typecast(const typecast_exprt &expr)
 {
-  assert(expr.operands().size()==1);
-
-  if(expr.op0().type().id()==ID_range)
+  if(expr.op().type().id() == ID_range)
   {
-    mp_integer from=string2integer(expr.op0().type().get_string(ID_from));
-    mp_integer to=string2integer(expr.op0().type().get_string(ID_to));
+    mp_integer from=string2integer(expr.op().type().get_string(ID_from));
```
Comment (peterschrammel) [inline src/solvers/flattening/boolbv_typecast.cpp]: clang-format

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 1fde831209d5a0f3f104d61b0d29813155b37635).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85546545

File: src/solvers/flattening/boolbv_quantifier.cpp [hunk 4]
```diff
@@ -140,11 +138,14 @@ exprt get_quantifier_var_max(
...
 {
-  if(!(expr.id()==ID_forall || expr.id()==ID_exists))
-    return true;
+  PRECONDITION(expr.id() == ID_forall || expr.id() == ID_exists);
 
-  assert(expr.operands().size()==2);
-  assert(expr.op0().id()==ID_symbol);
+  DATA_INVARIANT(
+    expr.operands().size() == 2,
+    "quantifier expressions shall have two operands");
```
Comment (tautschnig) [inline src/solvers/flattening/boolbv_quantifier.cpp]: I believe we have `to_quantifier_expr` by now, and that method should actually take a `quantifier_exprt`.

File: src/solvers/flattening/boolbv_quantifier.cpp [hunk 5]
```diff
@@ -203,6 +204,8 @@ bool instantiate_quantifier(exprt &expr,
 
 literalt boolbvt::convert_quantifier(const exprt &src)
 {
+  PRECONDITION(src.id() == ID_forall || src.id() == ID_exists);
```
Comment (tautschnig) [inline src/solvers/flattening/boolbv_quantifier.cpp]: The method should take a `quantifier_exprt`

File: src/solvers/flattening/boolbv_struct.cpp [hunk 6]
```diff
@@ -19,53 +19,53 @@ bvt boolbvt::convert_struct(const struct_exprt &expr)
...
+    expr.operands().size() == components.size(),
+    "number of operands of a struct expression shall equal the number of"
+    "components as indicated by its type",
+    expr.find_source_location());
 
   bvt bv;
   bv.resize(width);
 
-  std::size_t offset=0;
+  //std::size_t offset=0;
```
Comment (tautschnig) [inline src/solvers/flattening/boolbv_struct.cpp]: Delete?

File: src/solvers/flattening/boolbv_typecast.cpp [hunk 7]
```diff
@@ -590,20 +605,18 @@ bool boolbvt::type_conversion(
 /// conversion from bitvector types to boolean
 literalt boolbvt::convert_typecast(const typecast_exprt &expr)
 {
-  assert(expr.operands().size()==1);
-
-  if(expr.op0().type().id()==ID_range)
+  if(expr.op().type().id() == ID_range)
```
Comment (tautschnig) [inline src/solvers/flattening/boolbv_typecast.cpp:608]: Unnecessary change?

Comment (danpoe) [reply src/solvers/flattening/boolbv_typecast.cpp:608]: `op0()` -> `op()`

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: f971997ef4f9fd48a05ef9bb0beda06371a11d7c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85672302

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 213bfeb30bdbefeba78fb0580f8921d0587cdefd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85784112

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 44503dd06cfa6f08826298d058cd0115365e8384).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85871012
